File: android/GreenReadAR/app/src/main/kotlin/com/greenreadar/ar/GreenReadSession.kt
// GreenReadSession.kt
// GreenRead AR — Session Interface (Android)
// Port of GreenReadSession.swift — PRD §10.2-B
//
// Abstracts the AR session so that BallPhysicsEngine and SlopeAnalyzer
// depend only on this interface, enabling future alternative implementations
// (e.g. ARCore vs. test mock vs. replay session).

package com.greenreadar.ar

/**
 * 2D position on the green surface (x, z in world space).
 */
data class Float2(val x: Float, val y: Float)

/**
 * 3D position or vector in world space.
 */
data class Float3(val x: Float, val y: Float, val z: Float) {
    operator fun plus(other: Float3) = Float3(x + other.x, y + other.y, z + other.z)
    operator fun minus(other: Float3) = Float3(x - other.x, y - other.y, z - other.z)
    operator fun times(scalar: Float) = Float3(x * scalar, y * scalar, z * scalar)
    operator fun div(scalar: Float) = Float3(x / scalar, y / scalar, z / scalar)

    fun length(): Float = kotlin.math.sqrt(x * x + y * y + z * z)
    fun normalized(): Float3 {
        val len = length()
        return if (len > 0.0001f) Float3(x / len, y / len, z / len) else Float3(0f, 0f, 0f)
    }

    companion object {
        val zero = Float3(0f, 0f, 0f)
        fun dot(a: Float3, b: Float3): Float = a.x * b.x + a.y * b.y + a.z * b.z
        fun cross(a: Float3, b: Float3): Float3 = Float3(
            a.y * b.z - a.z * b.y,
            a.z * b.x - a.x * b.z,
            a.x * b.y - a.y * b.x
        )
    }
}

/**
 * Slope information for a point on the green.
 * Port of SlopeInfo struct from Swift.
 */
data class SlopeInfo(
    val degrees: Float,        // Angle from horizontal
    val percent: Float,        // Gradient as percentage
    val direction: String,     // Compass direction: "↑ N", "↘ SE", etc.
    val fallLine: Float3,      // Normalized fall direction vector
    val normal: Float3         // Surface normal
) {
    /** Color classification for the slope grid — PRD §5.3 */
    val colorCategory: SlopeColorCategory
        get() = when {
            degrees < 0.8f -> SlopeColorCategory.FLAT
            degrees < 2.5f -> if (fallLine.z > 0.1f) SlopeColorCategory.MODERATE_DOWNHILL
                              else SlopeColorCategory.MODERATE_UPHILL
            else -> if (fallLine.z > 0.1f) SlopeColorCategory.STEEP_DOWNHILL
                    else SlopeColorCategory.STEEP_UPHILL
        }
}

enum class SlopeColorCategory {
    FLAT,               // Green (#22C55E)
    MODERATE_DOWNHILL,  // Transitioning to blue
    MODERATE_UPHILL,    // Transitioning to red
    STEEP_DOWNHILL,     // Blue (#3B82F6)
    STEEP_UPHILL        // Red (#EF4444)
}

/**
 * Result from a completed ball roll simulation.
 * Port of BallRollResult struct from Swift.
 */
data class BallRollResult(
    val finalPosition: Float3,
    val trailPositions: List<Float3>,
    val totalDistance: Float,
    val maxBreak: Float,               // cm
    val breakDirection: BreakDirection,
    val avgSlopeAlongPath: Float,      // degrees
    val elevationChange: Float         // meters
) {
    enum class BreakDirection(val displayName: String) {
        LEFT("Vänster"),
        RIGHT("Höger"),
        STRAIGHT("Rak putt")
    }
}

/**
 * GreenReadSession — Core interface for AR session abstraction.
 * Port of GreenReadSession protocol from Swift — PRD §10.2-B.
 *
 * BallPhysicsEngine and SlopeAnalyzer depend ONLY on this interface,
 * not on the concrete ARCoreSession implementation.
 */
interface GreenReadSession {
    // Session Lifecycle
    fun startSession()
    fun pauseSession()
    fun stopSession()

    // Mesh & Scanning
    val scanProgress: Float
    val isGreenDetected: Boolean

    fun heightAt(position: Float2): Float?
    fun normalAt(position: Float2): Float3?
    fun slopeAt(position: Float2): SlopeInfo?

    // Entity Placement
    fun placeHole(at: Float3)
    fun placeBall(at: Float3)
    fun clearEntities()

    // Ball Simulation
    fun startBallRoll(stimpmeter: Float)
    val isBallRolling: Boolean

    // Rendering
    fun updateSlopeGrid()
    fun setBreakCurveVisible(visible: Boolean)
    fun configureBallTrail(fadeAfterSeconds: Double)

    // Callbacks
    var onGreenDetected: (() -> Unit)?
    var onBallStopped: ((BallRollResult) -> Unit)?
    var onSlopeUpdated: ((SlopeInfo) -> Unit)?
}

File: swift/GreenReadAR/AR/GreenReadSession.swift
// GreenReadSession.swift
// GreenRead AR — Session Protocol
// PRD §10.2-B: Abstracts rendering target for ARView (iPhone) vs ImmersiveSpace (visionOS)

import Foundation
import simd

// MARK: - GreenReadSession Protocol
/// Abstraction layer that enables the same scanning/analysis logic to run
/// on both iPhone (ARView) and visionOS (ImmersiveSpace/RealityView).
/// This is a KEY architectural decision per PRD §10.2-B for visionOS readiness.
protocol GreenReadSession: AnyObject {
    
    // MARK: - Session Lifecycle
    /// Start the AR scanning session
    func startSession()
    
    /// Pause the AR session (e.g., app backgrounded)
    func pauseSession()
    
    /// Stop and clean up the session
    func stopSession()
    
    // MARK: - Mesh & Scanning
    /// Current scan progress (0.0 – 1.0)
    var scanProgress: Float { get }
    
    /// Whether sufficient mesh data has been collected
    var isGreenDetected: Bool { get }
    
    /// Retrieve height at a world-space position on the green mesh
    func heightAt(position: SIMD2<Float>) -> Float?
    
    /// Retrieve surface normal at a world-space position
    func normalAt(position: SIMD2<Float>) -> SIMD3<Float>?
    
    /// Calculate slope information at a position
    func slopeAt(position: SIMD2<Float>) -> SlopeInfo?
    
    // MARK: - Entity Placement
    /// Place hole marker at world position
    func placeHole(at position: SIMD3<Float>)
    
    /// Place ball at world position
    func placeBall(at position: SIMD3<Float>)
    
    /// Remove all placed entities and reset
    func clearEntities()
    
    // MARK: - Ball Simulation
    /// Start ball roll simulation from current ball position toward hole
    func startBallRoll(stimpmeter: Float)
    
    /// Whether ball is currently rolling
    var isBallRolling: Bool { get }
    
    // MARK: - Rendering
    /// Update slope grid colors based on current mesh data
    func updateSlopeGrid()
    
    /// Show/hide the break curve overlay
    func setBreakCurveVisible(_ visible: Bool)
    
    /// Set trail visibility and fade duration
    func configureBallTrail(fadeAfterSeconds: TimeInterval)
    
    // MARK: - Callbacks
    var onGreenDetected: (() -> Void)? { get set }
    var onBallStopped: ((BallRollResult) -> Void)? { get set }
    var onSlopeUpdated: ((SlopeInfo) -> Void)? { get set }
}

// MARK: - Data Types

/// Slope information for a point on the green
struct SlopeInfo: Equatable {
    let degrees: Float        // Angle from horizontal
    let percent: Float        // Gradient as percentage
    let direction: String     // Compass direction: "↑ N", "↘ SE", etc.
    let fallLine: SIMD3<Float> // Normalized fall direction vector
    let normal: SIMD3<Float>   // Surface normal
    
    /// Color classification for the slope grid
    var colorCategory: SlopeColorCategory {
        if degrees < 0.8 {
            return .flat
        } else if degrees < 2.5 {
            return fallLine.z > 0.1 ? .moderateDownhill : .moderateUphill
        } else {
            return fallLine.z > 0.1 ? .steepDownhill : .steepUphill
        }
    }
}

enum SlopeColorCategory {
    case flat              // Green (#22C55E)
    case moderateDownhill  // Transitioning to blue
    case moderateUphill    // Transitioning to red
    case steepDownhill     // Blue (#3B82F6)
    case steepUphill       // Red (#EF4444)
}

/// Result from a completed ball roll simulation
struct BallRollResult {
    let finalPosition: SIMD3<Float>
    let trailPositions: [SIMD3<Float>]
    let totalDistance: Float
    let maxBreak: Float               // cm
    let breakDirection: BreakDirection
    let avgSlopeAlongPath: Float      // degrees
    let elevationChange: Float        // meters
    
    enum BreakDirection: String {
        case left = "Vänster"
        case right = "Höger"
        case straight = "Rak putt"
    }
}
